### PR TITLE
Distinguish between pseudo-class selectors and interpolation

### PR DESCRIPTION
**Pull request by kensternberg-authentik**
# Related Links:

Addresses [Issue 7: tree-sitter parses complex declaration as pseudo-class instead of CSS custom property](https://github.com/tree-sitter-grammars/tree-sitter-scss/issues/7).

# What

This commit places a guard clause around the scanner expression for detecting a pseudo-class selector colon.  The guard clause ensures that the character *prior to* any `{` symbol was not the start of interpolation.

I have also re-run the generate under 0.24.5, so the parser, headr files, and generated grammar have been updated to the latest & greatest version. A `tree-sitter.json` file has also been generated, and the corresponding tree-sitter clause removed from `package.json` (automatically, by tooling).

# Why

This declaration, which has been [added to the test cases](./test/corpus/declarations.txt).

```
div {
  --#{$x}--left: var(--#{$y}--right);
}
```

As written, when `PSEUDO_CLASS_SELECTOR_COLON` is a valid possible symbol, the scanner clause for pseudo-class selector detection passes if a `{` is detected before encountering a `}` or `;` symbol. The expression `#{$y}` in the test case meets that criteria, the lexer result is marked incorrectly, and the parser fails to parse the declaration correctly.

Adding a guard that the prior character seen was a `#` and that the `{` is not the start of a nested block identified by a pseudo-class but is, instead, part of an interpolation, fixes this issue.

# Test steps

1. Run the tests under the old clause and see this failure:

```
correct / expected / unexpected

  1. Interpolation on both sides of a declaration:

    (stylesheet
      (rule_set
        (selectors
          (tag_name))
        (block
          (declaration
            (ERROR)
            (property_name
              (identifier)
              (interpolation
                (variable))
              (identifier))
            (call_expression
              (function_name)
              (arguments
                (identifier)
                (interpolation
                  (variable))
                (identifier)))))))

```

2. Run it with the new clause and there is no more failure.  :-)  In fact, *all* of the tests now pass.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -49,16 +49,6 @@
     "parse": "tree-sitter parse",
     "test": "tree-sitter test"
   },
-  "tree-sitter": [
```
Comment (kensternberg-authentik) [inline package.json:52]: Change made by `tree-sitter generate` version 0.24.

File: src/scanner.c [hunk 2]
```diff
@@ -81,12 +93,25 @@ bool tree_sitter_scss_external_scanner_scan(void *payload, TSLexer *lexer, const
                 return false;
             }
             lexer->mark_end(lexer);
-            // We need a { to be a pseudo class selector, a ; indicates a property
-            while (lexer->lookahead != ';' && lexer->lookahead != '}' && !lexer->eof(lexer)) {
+            // We need a { to be a pseudo class selector, a ; indicates a
+            // property
+            while (lexer->lookahead != ';' && lexer->lookahead != '}' &&
```
Comment (kensternberg-authentik) [inline src/scanner.c:98]: This is where the actual fix begins; here I initialize a boolean to track the prior character, record it if necessary, and ignore the problematic `{` if `possible_interpolation` is still true.  I then remove the `possible_interpolation` if the follow-on character was not a problematic `{`.

Man, I hate going back and writing C again.

File: tree-sitter.json [hunk 3]
```diff
@@ -0,0 +1,37 @@
+{
```
Comment (kensternberg-authentik) [inline tree-sitter.json:1]: Added by `tree-sitter generate` with version 0.24.

File: src/scanner.c [hunk 4]
```diff
@@ -19,17 +19,24 @@ void tree_sitter_scss_external_scanner_destroy(void *payload) {}
 
 void tree_sitter_scss_external_scanner_reset(void *payload) {}
 
-unsigned tree_sitter_scss_external_scanner_serialize(void *payload, char *buffer) { return 0; }
+unsigned tree_sitter_scss_external_scanner_serialize(void *payload,
```
Comment (kensternberg-authentik) [inline src/scanner.c:22]: Most of this is `clang_format` and whatever the original code was formatted with disagreeing. The only syntactical change is marked below.

**Review by tanberry — APPROVED** [review]
Approving because I trust Ken with code and my life.

**Comment by kensternberg-authentik** [comment]
> Approving because I trust Ken with code and my life.

Hah! Thanks!  And although this works, trying it out on the Patternfly source code revealed a whole bunch of edge cases, so I'm going to have to withdraw it and try again. It`s *tree-sitter vs IBM*. "Whose cuisine will reign supreme!?"

**Pull request closed without merging**